server/service/src/sensor/berlinger.rs

```diff
@@ -0,0 +1,380 @@
+    RepositoryError, Sensor, SensorFilter, SensorRepository, SensorRow, SensorRowRepository,
+    SensorType, StorageConnection, TemperatureBreach, TemperatureBreachConfig,
+    TemperatureBreachConfigFilter, TemperatureBreachConfigRepository, TemperatureBreachConfigRow,
+    TemperatureBreachConfigRowRepository, TemperatureBreachFilter, TemperatureBreachRepository,
+    TemperatureBreachRow, TemperatureBreachRowRepository, TemperatureBreachRowType, TemperatureLog,
+    TemperatureLogFilter, TemperatureLogRepository, TemperatureLogRow, TemperatureLogRowRepository,
+};
+use util::uuid::uuid;
+
+use crate::{service_provider::ServiceContext, SingleRecordError};
+
+extern crate temperature_sensor;
+use temperature_sensor::*;
+
+pub fn get_breach_row_type(breach_type: &BreachType) -> TemperatureBreachRowType {
+    match breach_type {
+        BreachType::ColdConsecutive => TemperatureBreachRowType::ColdConsecutive,
+        BreachType::ColdCumulative => TemperatureBreachRowType::ColdCumulative,
+        BreachType::HotConsecutive => TemperatureBreachRowType::HotConsecutive,
+        BreachType::HotCumulative => TemperatureBreachRowType::HotCumulative,
+    }
+}
+
+pub fn get_matching_sensor_serial(
+    connection: &StorageConnection,
+    serial: &str,
+) -> Result<Vec<Sensor>, RepositoryError> {
+    SensorRepository::new(connection)
+        .query_by_filter(SensorFilter::new().serial(EqualFilter::equal_to(&serial)))
+}
+
+pub fn get_matching_sensor_log(
+    connection: &StorageConnection,
+    sensor_id: &str,
+    datetime: NaiveDateTime,
+) -> Result<Vec<TemperatureLog>, RepositoryError> {
+    let filter = TemperatureLogFilter::new()
+        .sensor(SensorFilter::new().id(EqualFilter::equal_to(sensor_id)))
+        .datetime(DatetimeFilter::equal_to(datetime));
+
+    TemperatureLogRepository::new(connection).query_by_filter(filter)
+}
+
+pub fn get_matching_sensor_breach_config(
+    connection: &StorageConnection,
+    description: &str,
+    breach_type: &TemperatureBreachRowType,
+) -> Result<Vec<TemperatureBreachConfig>, RepositoryError> {
+    let filter = TemperatureBreachConfigFilter::new()
+        .description(EqualFilter::equal_to(description))
+        .r#type(EqualFilter::equal_to_breach_type(&breach_type));
+
+    TemperatureBreachConfigRepository::new(connection).query_by_filter(filter)
+}
+
+pub fn get_matching_sensor_breach(
+    connection: &StorageConnection,
+    sensor_id: &str,
+    start_datetime: NaiveDateTime,
+    _end_datetime: NaiveDateTime,
+    breach_type: &TemperatureBreachRowType,
+) -> Result<Vec<TemperatureBreach>, RepositoryError> {
+    let filter = TemperatureBreachFilter::new()
+        .sensor(SensorFilter::new().id(EqualFilter::equal_to(sensor_id)))
+        .r#type(EqualFilter::equal_to_breach_type(&breach_type))
+        .start_datetime(DatetimeFilter::equal_to(start_datetime));
+
+    TemperatureBreachRepository::new(connection).query_by_filter(filter)
+}
+
+pub fn get_logs_for_sensor(
```

Comment:
I guess this ones is to get the length before and after, for logging ? Maybe best to do this to let the db do the work

  https://github.com/openmsupply/open-msupply/blob/cfbf37020a26ff900d9a3d3a305f8008d7ded9e4/server/repository/src/db_diesel/activity_log.rs#L48-L51

Reply:
My intention was that this fn (and similar ones for breaches and configs) would be called from elsewhere to return the logs etc. for a given sensor. I was just using them here for debugging while developing - can be deleted now I think (https://github.com/openmsupply/open-msupply/pull/2383/commits/927c969bc6845911b5c233945464e0c4f72e23ca)